File: api/linkextractor.rb
#!/usr/bin/env ruby
# encoding: utf-8

require "sinatra"
require "open-uri"
require "uri"
require "nokogiri"
require "json"
require "redis"

set :protection, :except=>:path_traversal

use_cache = ENV.fetch("USE_CACHE", "true") == "true"

redis = nil
if use_cache
  redis = Redis.new(url: ENV["REDIS_URL"] || "redis://localhost:6379")
  puts "[Ruby API] Redis conectado: #{ENV['REDIS_URL'] || 'redis://localhost:6379'}"
else
  puts "[Ruby API] Iniciando SEM cache."
end

Dir.mkdir("logs") unless Dir.exist?("logs")
cache_log = File.new("logs/extraction.log", "a")

get "/" do
  "Usage: http://<hostname>[:<prt>]/api/<url>"
end

get "/api/*" do
  url = [params['splat'].first, request.query_string].reject(&:empty?).join("?")
  cache_status = "BYPASS"
  jsonlinks = nil

  if use_cache && redis
    cache_status = "HIT"
    jsonlinks = redis.get(url)
    cache_status = "MISS" if jsonlinks.nil?
  end

  if jsonlinks.nil?
    jsonlinks = JSON.pretty_generate(extract_links(url))
    redis.set(url, jsonlinks) if use_cache && redis
  end

  cache_log.puts "#{Time.now.to_i}\t#{cache_status}\t#{url}"
  cache_log.flush

  status 200
  headers "content-type" => "application/json"
  body jsonlinks
end

def extract_links(url)
  links = []
  doc = Nokogiri::HTML(open(url))
  doc.css("a").each do |link|
    text = link.text.strip.split.join(" ")
    begin
      links.push({
        text: text.empty? ? "[IMG]" : text,
        href: URI.join(url, link["href"])
      })
    rescue
    end
  end
  links
end

File: api-python/linkextractor.py
#!/usr/bin/env python3

from flask import Flask, Response
from bs4 import BeautifulSoup
from urllib.request import urlopen, Request
from urllib.parse import urljoin
import redis
import json
import os

app = Flask(__name__)

USE_CACHE = os.environ.get("USE_CACHE", "true").lower() == "true"
redis_client = None

if USE_CACHE:
    redis_url = os.environ.get("REDIS_URL", "redis://localhost:6379")
    try:
        redis_client = redis.StrictRedis.from_url(redis_url, socket_connect_timeout=2)
        redis_client.ping()
        print(f"[Python API] Redis conectado: {redis_url}", flush=True)
    except Exception as e:
        print(f"[Python API] Redis indisponível: {e}. Cache desativado.", flush=True)
        redis_client = None
        USE_CACHE = False

if not USE_CACHE:
    print("[Python API] Iniciando SEM cache.", flush=True)


def extract_links(url):
    headers = {"User-Agent": "Mozilla/5.0 (compatible; LinkExtractor/1.0)"}
    req = Request(url, headers=headers)
    response = urlopen(req, timeout=10)
    html = response.read()
    soup = BeautifulSoup(html, "lxml")
    links = []
    for a_tag in soup.find_all("a", href=True):
        href = a_tag.get("href", "").strip()
        text = a_tag.get_text(strip=True) or "[IMG]"
        try:
            full_url = urljoin(url, href)
            links.append({"text": text, "href": full_url})
        except Exception:
            pass
    return links


@app.route("/")
def index():
    return "Usage: http://<hostname>[:<prt>]/api/<url>"


@app.route("/api/<path:url>")
def api(url):
    if USE_CACHE and redis_client:
        cached = redis_client.get(url)
        if cached:
            return Response(cached, status=200, mimetype="application/json")

    links = extract_links(url)
    json_links = json.dumps(links, indent=2, ensure_ascii=False)

    if USE_CACHE and redis_client:
        redis_client.set(url, json_links)

    return Response(json_links, status=200, mimetype="application/json")


if __name__ == "__main__":
    app.run(host="0.0.0.0", port=5000, debug=False)
